code/messaging/src/phone/java/com/adobe/marketing/mobile/MessagingInternal.java

```diff
@@ -471,6 +502,221 @@ private static void addXDMData(final EventData eventData, final Map<String, Obje
+        offersIdentifiers.put(MessagingConstant.EventDataKeys.Offers.ACTIVITY_ID, Variant.fromString(POC_ACTIVITY_ID_MULTI));
+        offersIdentifiers.put(MessagingConstant.EventDataKeys.Offers.PLACEMENT_ID, Variant.fromString(POC_PLACEMENT_ID_MULTI));
+        decisionScopes.add(offersIdentifiers);
+
+        final HashMap<String, Object> eventData = new HashMap<>();
+        eventData.put(MessagingConstant.EventDataKeys.Offers.TYPE, MessagingConstant.EventDataKeys.Offers.PREFETCH);
+        eventData.put(MessagingConstant.EventDataKeys.Offers.DECISION_SCOPES, decisionScopes);
+
+        final Event messageFetchEvent = new Event.Builder(MessagingConstant.EventName.MESSAGING_REFRESH_IAM, MessagingConstant.EventType.OFFERS, EventSource.REQUEST_CONTENT.getName())
+                .setEventData(eventData)
+                .build();
+
+        // send event
+        MobileCore.dispatchEvent(messageFetchEvent, new ExtensionErrorCallback<ExtensionError>() {
+            @Override
+            public void error(ExtensionError extensionError) {
+                Log.error(LOG_TAG, "%s - Error in dispatching event for fetching messages from Offers.", SELF_TAG);
+            }
+        });
+    }
+
+    /**
+     * Converts the rules json present in the Edge response event payload into a
+     * list of rules then loads them in the rules engine.
+     * @param edgeResponseEvent An Edge {@link Event} containing a personalization decision payload retrieved from Offers.
+     */
+    private void handleOfferNotificationPayload(final Event edgeResponseEvent) {
+        final JSONObject payload = (JSONObject) edgeResponseEvent.getEventData().get(MessagingConstant.EventDataKeys.Offers.PAYLOAD);
+        if (payload == null || payload.length() == 0) {
+            Log.warning(LOG_TAG, "handleOfferNotification - Aborting handling of the Offers IAM payload because it is null or empty.");
+            return;
+        }
+
+        String rulesJsonString = null;
+        try {
+            final JSONObject activity = payload.getJSONObject(MessagingConstant.EventDataKeys.Offers.ACTIVITY);
+            final JSONObject placement = payload.getJSONObject(MessagingConstant.EventDataKeys.Offers.PLACEMENT);
+            final String offerActivityId = (String) activity.get(MessagingConstant.EventDataKeys.Offers.ID);
+            final String offerPlacementId = (String) placement.get(MessagingConstant.EventDataKeys.Offers.ID);
+            if (!offerActivityId.equals(POC_ACTIVITY_ID_MULTI) || !offerPlacementId.equals(POC_PLACEMENT_ID_MULTI)) {
+                Log.warning(LOG_TAG, "handleOfferNotification - ignoring Offers IAM payload, the expected offer activity id or placement id is missing. ");
+                return;
+            }
+            final JSONObject items = payload.getJSONArray(MessagingConstant.EventDataKeys.Offers.ITEMS).getJSONObject(0);
```

Comment:
we should be looping through this `ITEMS` array and getting the json for each one.  it's expected that this array contains all of the in-app messages designated for this user.

Reply:
i see, will update this as well.